myia/composite.py

```diff
@@ -561,6 +579,40 @@ def _array_zero(xs):
 )
 
 
+class IsCompare(MetaGraph):
+    """Implementation of Is Compare (i.e. 'is' and 'is not')."""
+    def __init__(self, do_not=False):
+        self.do_not = do_not
+
+    def normalize_args(self, args):
+        return tuple(broaden(a, None) for a in args)
+
+    def generate_graph(self, args):
+        """Generate Boolean Compare if False and
+        Equal_To (and not Equal_To) compare if True.
+
+        """
+        assert len(args) == 2
+        a, b = args
+        g = Graph()
+        g.debug.name = "is_not" if self.do_not else "is_"
+        g.set_flags('core', 'reference')
+        pa = g.add_parameter()
+        pb = g.add_parameter()
+
+        if a != b:
+            g.return_ = g.apply(P.return_, self.do_not)
+        else:
+            cmp_fn = g.apply(P.getattr, pa,
+                             "__ne__" if self.do_not else "__eq__")
```

Comment:
`is` and `is not` should only compare to Bool and None. On any other type, they do identity comparison in Python, and we probably don't want to support it. `is` and `is not` do _not_ call `__eq__` or `__ne__`.

To be clear, at least one of the arguments should be a bool or None, and then we can check if the other is also None or the right boolean. If both arguments are scalars or arrays or anything else, there should be a compile time type error.

Reply:
I think it's ok to make is equivalent to `__eq__` if the types are the same.  Because we are functional this is essentially the same.

Reply:
To do so would make Myia's behavior deviate from Python's. If you write `a is b` instead of `a == b` in Python, you mean to check whether `a` and `b` are the same object, and it would be problematic if Myia didn't do that.

For example, if a user decides to do something like `if x is layers[0].weights: ...` to do something on the first weights layer, running `x == layers[0].weights` instead is going to be both inefficient and potentially erroneous if some other parameter has the same values for some reason (e.g. if they all start out at 0). I think it's best to forbid it with an informative error message.